src/Group.jl

```diff
@@ -0,0 +1,319 @@
+
+"""
+    GroupManifold{M<:Manifold,O<:AbstractGroupOperation} <: AbstractGroupManifold{O}
+
+Decorator for a smooth manifold that equips the manifold with a group
+operation, thus making it a Lie group. See [`AbstractGroupManifold`](@ref) for
+more details.
+
+# Constructor
+
+    GroupManifold(manifold, op)
+"""
+struct GroupManifold{M<:Manifold,O} <: AbstractGroupManifold{O}
+    manifold::M
+    op::O
+end
+
+@traitimpl IsDecoratorManifold{GroupManifold}
+
+struct LeftInvariantCanonicalMetric <: RiemannianMetric end
+struct RightInvariantCanonicalMetric <: RiemannianMetric end
+
+@traitimpl HasMetric{GroupManifold,LeftInvariantCanonicalMetric}
+
+"""
+    IsMatrixGroup
+
+A `Trait` to indicate that a group's default representation is a matrix
+representation. For groups with a `MultiplicationOperation`, this indicates
+that the group is a subgroup of the General Linear Group.
+"""
+@traitdef IsMatrixGroup{G}
+
+@doc doc"""
+    Identity
+
+The identity element of any group.
+"""
+struct Identity{G<:AbstractGroupManifold}
+    group::G
+end
+
+(e::Identity)(x) = identity(e.group, x)
+
+@doc doc"""
+    inv(G::AbstractGroupManifold, x)
+
+Inverse $x^{-1}$ of an element $x$, such that
+$x \cdot x^{-1} = x^{-1} \cdot x = e$.
+"""
+function inv(G::AbstractGroupManifold, x)
+    error("inv not implemented on $(typeof(G)) for point $(typeof(x))")
+end
+
+inv(::AbstractGroupManifold, e::Identity) = e
+
+@doc doc"""
+    identity(G::AbstractGroupManifold, x)
+
+Identity element $e$, such that for any element $x$,
+$x \cdot e = e \cdot x = x$. The returned element is of a similar type to `x`.
+"""
+function identity(G::AbstractGroupManifold, x)
+    error("identity not implemented on $(typeof(G)) for point $(typeof(x))")
+end
+
+@doc doc"""
+    left_action(G::AbstractGroupManifold, x, p)
+
+For an element $p$ of some set, compute the left action of group element $x$
+on $p$, i.e. $x \cdot p$.
+"""
+function left_action(G::AbstractGroupManifold, x, p)
+    error("left_action not implemented on $(typeof(G)) for point $(typeof(x)) and object $(typeof(p))")
+end
+
+@doc doc"""
+    right_action(G::AbstractGroupManifold, x, p)
+
+For an element $p$ of some set, compute the right action of group element $x$
+on $p$, i.e. $p \cdot x$. The default right action is constructed from the left
+action, i.e. $p \cdot x = x^{-1} \cdot p$.
+"""
+function right_action(G::AbstractGroupManifold, x, p)
+    return left_action(G, inv(G, x), p)
+end
+
+# Adapted from `afoldl` in `operators.jl` in Julia base.
+# expand recursively up to a point, then switch to a loop.
+group_afoldl(op, G, a) = a
+group_afoldl(op, G, a, b) = op(G,a,b)
+group_afoldl(op, G, a, b, c...) = group_afoldl(op, G, op(G, a, b), c...)
+
+function group_afoldl(op,G,a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,qs...)
+    y = op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,op(G,a,b),c),d),e),f),g),h),i),j),k),l),m),n),o),p)
+    return op(G, y, reduce((x, y)->op(G, x, y), qs))
+end
```

Comment:
Can you explain this a little, it looks confusing with so many operators, what happens for example if you call it with more than that?

Reply:
Yeah this is derived from `operators.jl` in Julia base. It's used so expressions like `+(a,b,c,d,e,f)` get efficiently inlined when there are up to 16 elements. In base, the elements in `qs` are iterated through. Here, we use `reduce`, which I think will be more efficient (still need to benchmark this specific choice).